src/paths/steiner.cpp

```diff
@@ -0,0 +1,586 @@
+    IGRAPH_CHECK(generateSubsets(&steiner_terminals_copy, igraph_vector_int_size(&steiner_terminals_copy), no_of_nodes, subsetMap, allSubsets));
+
+    int steiner_terminals_copy_size = igraph_vector_int_size(&steiner_terminals_copy);
+    for (igraph_integer_t m = 2; m <= steiner_terminals_copy_size; m++) {
+        for (igraph_integer_t i = 0; i < (igraph_integer_t)allSubsets.size(); i++) {
+            auto it = allSubsets.begin();
+            std::advance(it, i);
+            int_set D = *it;
+            igraph_integer_t indexOfSubsetD;
+            indexOfSubsetD = fetchIndexofMapofSets(D, subsetMap);
+
+            IGRAPH_ALLOW_INTERRUPTION();
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                MATRIX(dp_cache, indexOfSubsetD, j) = IGRAPH_INFINITY;
+            }
+
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                igraph_real_t distance1 = IGRAPH_INFINITY;
+
+                for (auto E :  D) {
+
+                    if (E != j) {
+                        igraph_real_t distanceEJ = MATRIX(distance, E, j);
+
+                        int_set DMinusE = D;
+
+                        /*
+                         *  A set with Singleton value E removed from subset D
+                         */
+                        for (auto item :  DMinusE) {
+                            if (item == E) {
+                                DMinusE.erase(item);
+                                break;
+                            }
+                        }
+
+                        igraph_integer_t indexOfSubsetDMinusE;
+                        if (DMinusE.size() == 1) {
+                            indexOfSubsetDMinusE = *DMinusE.begin();
+                        } else {
+                            indexOfSubsetDMinusE = fetchIndexofMapofSets(DMinusE, subsetMap);
+                        }
+
+                        if ((distanceEJ + MATRIX(dp_cache, indexOfSubsetDMinusE, j)) < distance1) {
+                            distance1 = distanceEJ + (MATRIX(dp_cache, indexOfSubsetDMinusE, j));
+                        }
+                    }
+                }
+
+                for (igraph_integer_t k = 0; k < no_of_nodes; k++) {
+                    MATRIX(dp_cache, indexOfSubsetD, k) = std::min(MATRIX(dp_cache, indexOfSubsetD, k), MATRIX(distance, k, j) + distance1);
+                }
+            }
+        }
+    }
+
+    igraph_real_t distance2 = IGRAPH_INFINITY;
+
+    for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+        igraph_real_t distance1 = IGRAPH_INFINITY;
+        IGRAPH_ALLOW_INTERRUPTION();
+        for (igraph_integer_t subset_C_iterator = 0; subset_C_iterator < no_of_terminals; subset_C_iterator++) {
+            igraph_integer_t F = VECTOR(steiner_terminals_copy)[subset_C_iterator];
+            igraph_real_t distanceFJ = MATRIX(distance, F, j);
+
+            int_set CMinusF;
+
+            for (igraph_integer_t k = 0; k < no_of_terminals; k++) {
```

Comment:
Isn't this all way easier if you make C a set too, instead of a vector? Although I'm not sure about the performance.

Reply:
`typedef std::set<igraph_integer_t> int_set;` C and CMinusF are both of the type int_set. I maybe have misunderstood but it is already a set

Reply:
I was calling `steiner_terminals_copy` `C` because you use `subset_C_iterator` to iterate over it, and copying all of its elements except `F` results in CMinusF.

`steiner_terminals_copy` is a vector, and therefore you can't just take it and do `.erase(F)` like you could with `DMinusE`.